include/openrave/openravecontainer.h

```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        if (invalidIndex >= 0) {
+            // reclaim existing invalid data
+            _vNamedDatas[invalidIndex].Set(nameId, data);
+
+            // update begin and end index
+            if (invalidIndex < _beginValidElementsIndex) {
+                _beginValidElementsIndex = invalidIndex;
+            }
+            if (invalidIndex + 1 > _endValidElementsIndex) {
+                _endValidElementsIndex = invalidIndex + 1;
+            }
+        }
+        else {
+            // cannot reclaim, insert new data.
+            _vNamedDatas.push_back(NamedData<DataType>(nameId, data));
+            _endValidElementsIndex++;
+        }
+        _numValidElements++;
+        OPENRAVE_ASSERT_OP(_numValidElements, <= ,(_endValidElementsIndex - _beginValidElementsIndex));
+    }
+
+    /// \brief Iterator class to be used for Map class
+    class Iterator
+    {
+    public:
+        Iterator() = delete;
+        Iterator(const std::vector<NamedData<DataType>>& data, size_t dataIndex, size_t endIndex)
+            : _data(data), _dataIndex(dataIndex), _endIndex(endIndex)
+        {
+        }
+
+        /// \brief Pre-increment
+        /// iterate to next valid data or to end index. Even when iterator is beyond end index, iterates to the next index once.
+        Iterator& operator++()
+        {
+            do {
+                ++_dataIndex;
+            } while (_dataIndex < _endIndex && !_data[_dataIndex].IsValid());
+
+            return *this;
+        }
+
+        /// \brief compares for equality
+        /// \param other the other object to compare against
+        bool operator==(const Iterator& other) const { return _dataIndex == other._dataIndex && this == &other; }
+
+        /// \brief compares for inequality
+        /// \param other the other object to compare against
+        bool operator!=(const Iterator& other) const { return !(*this == other); }
+
+        /// \brief dereference operator
+        const NamedData<DataType>& operator*() const
+        {
+            return _data[_dataIndex];
+        }
+    private:
+        const std::vector<NamedData<DataType>>& _data; ///< underlying data to iterate over
+        size_t _dataIndex; ///< current position
+        const size_t _endIndex; ///< end position
+    };
+
+    /// \brief gets iterator to the first element
+    /// \return iterator to first element
+    Iterator GetBegin() const
+    {
+        return Iterator(_vNamedDatas, _beginValidElementsIndex, _endValidElementsIndex);
+    }
+
+    /// \brief gets iterator to the element after last element
+    /// \return iterator to the element after last element
+    Iterator GetEnd() const
+    {
+        return Iterator(_vNamedDatas, _endValidElementsIndex, _endValidElementsIndex);
+    }
+
+private:
+    std::vector<NamedData<DataType>> _vNamedDatas; ///< Vector of elements. Vector is not sorted. For small size, faster to keep it unsorted and do brute-force search.
```

Comment:
Why not maintain this in binary order via std::_upper_bound? Given the low number of objects we deal with here, perhaps doesn't make a difference ;0/

Reply:
Yes, I was thinking that we may not benefit much from sorting given relatively small number of elements.

Also, because we are not actually removing elements from this vector in `Erase` to preserve memory, it is computationally expensive to keep it sorted. One insertion or erasing operation can trigger to manipulate many elements to keep this vector sorted.